lib/Echidna/ABI.hs

```diff
@@ -309,16 +304,18 @@ mutateAbiCall = traverse f
 -- @a@ from a GenDict, return a generator that takes an @a@ and either synthesizes new @b@s with the
 -- provided generator or uses the 'GenDict' dictionary (when available).
 genWithDict :: (Eq a, Hashable a, MonadRandom m)
-            => GenDict -> HashMap a [b] -> (a -> m b) -> a -> m b
+            => GenDict -> HashMap a (Set b) -> (a -> m b) -> a -> m b
 genWithDict genDict m g t = do
   r <- getRandom
-  let maybeValM = if genDict ^. pSynthA >= r then fromDict else pure Nothing
-      fromDict = uniformMay (M.lookupDefault [] t m)
+  let maybeValM = if genDict._pSynthA >= r then fromDict else pure Nothing
+      fromDict = case M.lookup t m of
+                   Nothing -> pure Nothing
+                   Just cs -> Just <$> rElem' cs
   fromMaybe <$> g t <*> maybeValM
 
 -- | Synthesize a random 'AbiValue' given its 'AbiType'. Requires a dictionary.
 genAbiValueM :: MonadRandom m => GenDict -> AbiType -> m AbiValue
-genAbiValueM genDict = genWithDict genDict (toList <$> genDict ^. constants) $ \case
+genAbiValueM genDict = genWithDict genDict genDict._constants $ \case
```

Comment:
Should this be faster than using `toList`, right?

Reply:
yes, picking random elements from Set is faster (O(log n)) than from a list (O(n))